include/cppkafka/utils/buffered_producer.h

```diff
@@ -187,22 +215,25 @@ void BufferedProducer<BufferType>::add_message(Builder builder) {
 template <typename BufferType>
 void BufferedProducer<BufferType>::produce(const MessageBuilder& builder) {
     produce_message(builder);
-    expected_acks_++;
 }
 
 template <typename BufferType>
 void BufferedProducer<BufferType>::flush() {
-    while (!messages_.empty()) {
+    size_t num_messages = messages_.size();
```

Comment:
I'll have a deeper look later and I don't sure if this would cause some issues with the expected ack number but I normally handle these sort of blocking operations like this:

```c++
// We'll lock and move the messages out. This makes pending_messages
// have all current messages and messages_ is left empty
QueueType pending_messages = [&]() {
    lock_guard<mutex> _(mutex_);
    return move(messages_);
}();
while (!pending_messages.empty()) {
    // same loop, no lock
}
```

This way you have no potentially blocking or long operations inside the critical section while achieving the same. Do you think this would work here?

Reply:
I though about a similar swap approach with a temp queue (like you use in `clear`) but from my experience it's more hassle as now you have to deal with potentially two queues. What if after the swap the produce fails? Say the failure callback the user has provided returns false, then you have to stop retrying. Now you have 2 queues. Do you merge them back? Next time you flush again you risk having to flush 2 queues...not worth it IMHO. I've done this in the past and the management overhead is too much.

Reply:
BTW, note there's only a single ack counter in my revision, since producers and the flush threads are both operating in parallel. The test cases I provided are pretty reliable.

Reply:
If production fails and the callback returns false then you just discard the message, right? That's how it works right now. If a message keeps failing for a particular reason (say the message size is larger than the max allowed in the config for that topic), then you want to discard it, otherwise you'll be stuck forever.

Since you're using a single counter that gets incremented on produce and decremented on ack (both on fail and success), the counter will have the right value. I think there's no need to handle 2 queues, unless I'm missing something,

Reply:
That's a good point and actually, `on_delivery_report` shouldn't throw. That's inside rdkafka's stack and that will probably break things if it happens. Not sure if we want to allow exceptions to even happen within the callback context, that sounds very dangerous.

Reply:
Actually this brings a good point, the failure callback is optional, if you don't supply it you will be stuck forever in case of an error. There should be some kind of failsafe mechanism OR force failure callback as part of the constructor.

Reply:
>That's a good point and actually, on_delivery_report shouldn't throw. That's inside rdkafka's stack and that will probably break things if it happens. Not sure if we want to allow exceptions to even happen within the callback context, that sounds very dangerous.

No, no...you missunderstood, it's the `produce_message` that throws.

Reply:
True. I think probably it makes sense to force the callback to be set. I don't think there's a sane strategy otherwise, given we can't really keep track of messages between produce calls so from cppkafka's perspective it's not easy to know how many times a single message has been re-produced so as to have some sort of limit on how many times this can happen.